### PR TITLE
Fix RangeError due to stale signature help tooltips

### PR DESCRIPTION
**Pull request by microbit-matt-hillsdon**
It now maps the position through so you see lagged tooltips if Pyright responses are slow rather than tooltips for positions that might not be in the document at all.

### CODE REVIEW

File: src/editor/codemirror/language-server/signatureHelp.ts [hunk 1]
```diff
@@ -99,27 +109,55 @@ export const signatureHelp = (
 ) => {
   const signatureHelpTooltipField = StateField.define<SignatureHelpState>({
     create: () => ({
+      pos: -1,
       result: null,
-      tooltip: null,
     }),
     update(state, tr) {
+      let pos = state.pos === -1 ? -1 : tr.changes.mapPos(state.pos);
+      console.log(state.pos, pos);
```
Comment (microbit-matt-hillsdon) [inline src/editor/codemirror/language-server/signatureHelp.ts:117]: Remove debug.

```suggestion
```

File: src/editor/codemirror/language-server/signatureHelp.ts [hunk 2]
```diff
@@ -99,27 +109,55 @@ export const signatureHelp = (
...
+      return {
+        pos,
+        result,
+      };
     },
-    provide: (f) => showTooltip.from(f, (val) => val.tooltip),
+    provide: (f) =>
+      showTooltip.from(f, (val) => {
+        const { result, pos } = val;
+        console.log(result, pos);
```
Comment (microbit-matt-hillsdon) [inline src/editor/codemirror/language-server/signatureHelp.ts:138]: Remove debug.

```suggestion
        console.log(result, pos);
```

Comment (microbit-matt-hillsdon) [reply src/editor/codemirror/language-server/signatureHelp.ts:138]: You see this as (125, 124) in the scenario that otherwise would be bug #1053

Comment (microbit-matt-hillsdon) [reply src/editor/codemirror/language-server/signatureHelp.ts:137]: Not sure why I can't commit this from the GH web UI. Will resolve and re-raise.

Comment (microbit-matt-hillsdon) [reply src/editor/codemirror/language-server/signatureHelp.ts:137]: Ah, I see, it was a no-op change. Edited to remove.

File: src/editor/codemirror/language-server/signatureHelp.ts [hunk 3]
```diff
@@ -72,17 +77,22 @@ const triggerSignatureHelpRequest = async (view: EditorView): Promise<void> => {
...
     view.dispatch({
-      effects: [setSignatureHelpEffect.of({ pos, result })],
+      effects: [setSignatureHelpResult.of(result)],
     });
   } catch (e) {
     logException(view.state, e, "signature-help");
     view.dispatch({
-      effects: [setSignatureHelpEffect.of({ pos, result: null })],
+      effects: [setSignatureHelpResult.of(null)],
     });
```
Comment (microbit-matt-hillsdon) [inline src/editor/codemirror/language-server/signatureHelp.ts:96]: The essence of the change is that we no longer assume this position is still valid when we hear back from Pyright.
Instead we stash it before the request and map it forward through subsequent changes as required.

**Comment by microbit-matt-hillsdon** [comment]
> The slightly buggy behaviour of having signature help on an existing function open, and then dragging over a new API snippet I mentioned previously causes errors on this branch.

Thanks, I can reproduce them.

File: src/editor/codemirror/language-server/signatureHelp.ts [hunk 4]
```diff
@@ -63,32 +68,42 @@ const signatureHelpToolTipBaseTheme = EditorView.baseTheme({
...
+  const client = state.facet(clientFacet)!;
+  const pos = state.selection.main.from;
   const params: SignatureHelpParams = {
     textDocument: { uri },
-    position: offsetToPosition(view.state.doc, pos),
+    position: offsetToPosition(state.doc, pos),
   };
   try {
+    // Must happen before other event handling that might dispatch more
+    // changes thatt invalidate our position.
```
Comment (microbit-matt-hillsdon) [inline src/editor/codemirror/language-server/signatureHelp.ts:84]:
```suggestion
    // changes that invalidate our position.
```

**Comment by microbit-matt-hillsdon** [comment]
I think this is now working. @microbit-robert, please break it again!

I want this to work without the fix of closing signature help as I think dnd actually exercises this code really well. However, that's still a good plan so I've raised #1055 and have a fix on #1056 that I'll rebase after this is in.

**Comment by microbit-matt-hillsdon** [comment]
I'm re-running CI (it passed) to build a bit more confidence in the e2e tests here.

**Comment by microbit-matt-hillsdon** [comment]
Looking again at this as I managed to trigger an error with my original repro.

**Comment by microbit-matt-hillsdon** [comment]
> Looking again at this as I managed to trigger an error with my original repro.

OK, I'm happy again 🤞

It's rebuilding now.

**Review by microbit-robert — APPROVED** [review]
Now looks good. I can occasionally see some lag in updating the tooltip as expected, but I can't produce any errors.

**Pull request merged**